File: sample/src/main/java/net/huray/omronsdk/ui/device_list/DeviceListActivity.java
package net.huray.omronsdk.ui.device_list;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.widget.ListView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import net.huray.omronsdk.R;
import net.huray.omronsdk.ui.register.OmronDeviceRegisterActivity;
import net.huray.omronsdk.ui.request_data.OmronTransferActivity;
import net.huray.omronsdk.utils.Const;

public class DeviceListActivity extends AppCompatActivity {
    public static final int REQUEST_CODE = 11;
    private final String permission = Manifest.permission.ACCESS_FINE_LOCATION;

    private DeviceListAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_device_list);

        requestPermission();
        initViews();
    }

    @Override
    protected void onResume() {
        super.onResume();
        initViews();
    }

    private void initViews() {
        adapter = new DeviceListAdapter(this);
        ListView listView = findViewById(R.id.lv_device_list);
        listView.setAdapter(adapter);
        listView.setOnItemClickListener((parent, view, position, id) -> moveScreen(position));
    }

    private void moveScreen(int position) {
        if (isPermissionGranted()) {
            moveToActivity(position);
            return;
        }

        requestPermission();
    }

    private void moveToActivity(int position) {
        if (adapter.getDeviceConnectionState(position)) {
            Intent intent = new Intent(this, OmronTransferActivity.class);
            intent.putExtra(Const.EXTRA_DEVICE_TYPE, adapter.getDeviceTypeNumber(position));
            startActivity(intent);
            return;
        }

        Intent intent = new Intent(this, OmronDeviceRegisterActivity.class);
        intent.putExtra(Const.EXTRA_DEVICE_TYPE, adapter.getDeviceTypeNumber(position));
        startActivity(intent);
    }

    private void requestPermission() {
        if (isPermissionGranted()) return;

        ActivityCompat.requestPermissions(this, new String[]{permission}, REQUEST_CODE);
    }

    private boolean isPermissionGranted() {
        return ContextCompat.checkSelfPermission(this, permission) ==
                PackageManager.PERMISSION_GRANTED;
    }
}

File: sample/src/main/java/net/huray/omronsdk/ui/device_list/DeviceListActivity.kt
package net.huray.omronsdk.ui.device_list

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import net.huray.omronsdk.databinding.ActivityDeviceListBinding
import net.huray.omronsdk.ui.register.DeviceRegisterActivity
import net.huray.omronsdk.ui.transfer.DeviceTransferActivity
import net.huray.omronsdk.utils.Const

class DeviceListActivity : AppCompatActivity(), DeviceItemClickListener {
    private val permission = Manifest.permission.ACCESS_FINE_LOCATION

    private var adapter: DeviceListAdapter = DeviceListAdapter(this)

    private lateinit var binding: ActivityDeviceListBinding

    private val isPermissionGranted: Boolean
        get() = ContextCompat.checkSelfPermission(this, permission) ==
                PackageManager.PERMISSION_GRANTED

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityDeviceListBinding.inflate(layoutInflater)
        setContentView(binding.root)
        binding.lifecycleOwner = this
        binding.rvDeviceList.adapter = adapter

        requestPermission()
    }

    override fun onResume() {
        adapter.initDeviceList()
        super.onResume()
    }

    override fun onItemClicked(isConnected: Boolean, deviceId: Int) {
        moveScreen(isConnected, deviceId)
    }

    private fun moveScreen(isConnected: Boolean, deviceId: Int) {
        if (isPermissionGranted) {
            moveToActivity(isConnected, deviceId)
            return
        }
        requestPermission()
    }

    private fun moveToActivity(isConnected: Boolean, deviceId: Int) {
        if (isConnected) {
            val intent = Intent(this, DeviceTransferActivity::class.java)
            intent.putExtra(Const.EXTRA_DEVICE_TYPE, deviceId)
            startActivity(intent)
            return
        }

        val intent = Intent(this, DeviceRegisterActivity::class.java)
        intent.putExtra(Const.EXTRA_DEVICE_TYPE, deviceId)
        startActivity(intent)
    }

    private fun requestPermission() {
        if (isPermissionGranted) return
        ActivityCompat.requestPermissions(this, arrayOf(permission), REQUEST_CODE)
    }

    companion object {
        const val REQUEST_CODE = 11
    }
}